client: don't hang when MDS session is evicted

rishabh-d-dave opened the pull request:
This PR adds the code that makes an evicted client not hang on a MDS request.

rishabh-d-dave commented:
There is an issue with the code in this PR though - the client doesn't hang and also unmounts (if it has to) but ceph-fuse doesn't terminate. The process stays alive in the background. The issues seems to be with the client object because commenting off `client->shutdown()` and  `delete client; client = NULL` makes it terminate immediately. Oddly, `Client::shutdown()` is being called 5 times (of which it hangs at the 5th time) and I see it being called only once in the code. Output from debug code

```
ceph, Client.cc: Client::shutdown() starts
ceph, Client.cc: Client::shutdown() returning
ceph, Client.cc: Client::shutdown() starts
ceph, Client.cc: Client::shutdown() returning
ceph, Client.cc: Client::shutdown() starts
ceph, Client.cc: Client::shutdown() returning
ceph, Client.cc: Client::shutdown() starts
ceph, Client.cc: Client::shutdown() returning
ceph, Client.cc: Client::shutdown() starts
```

How do I figure out how/where were the calls made?

rishabh-d-dave commented on src/client/Client.cc:
```diff
@@ -6011,6 +6012,9 @@ void Client::tick()
   }
 
+  if (mounted && !mds_requests.empty() && now - curr_req_stamp > 120)
+    objecter->get_latest_version(0, curr_req_stamp, NULL);
```
While the patch fixes the issue, there is, at least, one technical issue: the second parameter is incorrect. It is supposed to be of the type `epoch_t`. Client.cc has only 2 variables that are supposed to hold epoch values: `cap_epoch_barrier` and `local_osd_epoch`. But both of them hold the value `0` when the client has hanged. So passing any of them wouldn't update client's OSD map.

batrick commented:
@rishabh-d-dave, please also look at `qa/tasks/cephfs/test_misc.py` and add a test case for the bug. i.e. a test case where (ceph-fuse) client cannot unmount after eviction which your patch fixes.

I'll send you details on how to do testing on a vstart cluster by email.

rishabh-d-dave commented on src/client/Client.cc:
```diff
@@ -6011,6 +6013,16 @@ void Client::tick()
     }
   }
 
+  {
```
After using right data types, this block looked overlong and unreadable. So, I've declared a couple of variables to express the logic succinctly and, then, added a block-scope since these variables won't be required for long.

batrick replied:
sure

rishabh-d-dave commented:
Retesting the patch since failure was apparently due to something else -
```
[  7%] Built target rocksdb_ext
CMakeFiles/Makefile2:483: recipe for target 'CMakeFiles/tests.dir/rule' failed
make[1]: *** [CMakeFiles/tests.dir/rule] Error 2
Makefile:264: recipe for target 'tests' failed
make: *** [tests] Error 2
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Executing post build scripts.
[ceph-pull-requests] $ /bin/sh -xe /tmp/jenkins8797732219848046472.sh
+ sudo reboot
FATAL: command execution failed
java.io.EOFException

rishabh-d-dave commented:
retest this please

batrick commented:
@rishabh-d-dave rebase this please so `make check` succeeds.

batrick commented:
Also, how is writing the test going?

rishabh-d-dave commented:
> Also, how is writing the test going?

Sorry about the delay. Had to fix some of my own patches for Manila.

rishabh-d-dave commented on qa/tasks/cephfs/test_misc.py:
```diff
@@ -10,6 +10,7 @@
 
 class TestMisc(CephFSTestCase):
     CLIENTS_REQUIRED = 2
+    MDSS_REQUIRED = 2
```
Is there are way that number of MDSs can be changed for only one test case rather than the entire group of test cases? Because this changes the environment for all the test cases which is incorrect in my opinion.

But, on other hand, it shouldn't affect the other test cases since the extra MDS(s) won't be active on their filesystem(s) until max_mds is explicitly set to 2

rishabh-d-dave commented on qa/tasks/cephfs/test_misc.py:
```diff
@@ -129,6 +130,29 @@ def test_evict_client(self):
+                                            'id={0}'.format(client_id))
+
+        # Test: Instead of hanging, client should unmount.
+        self.mount_a.umount()
```
This line reveals an issue: the method `umount()` uses `fusermount` to unmount the filesystem. The client unmounts successfully but ceph-fuse remains alive in the background. This makes the method `umount_wait()` (on the next line) wait until the timeout and, as a result, the test fails. I see the same behaviour when I do it manually.

Contradictorily, using umount command leads to both - CephFS vacating the mount point as well as exiting ceph-fuse process. So, must this mean a bug on the Ceph side?

batrick replied:
> Contradictorily, using umount command leads to both - CephFS vacating the mount point as well as exiting ceph-fuse process. So, must this mean a bug on the Ceph side?

That's interesting. @ukernel @jtlayton do you have an explanation why `umount` would cause ceph-fuse to exit when `fusermount -u` does not?

batrick commented:
@rishabh-d-dave how's this going?

rishabh-d-dave commented:
> @rishabh-d-dave how's this going?

@batrick The test results have been inconsistent. Now I get the same result with `umount` as well as `fusermount` -- ceph-fuse exits but with the return value -6. I get the following backtraces from gdb and the test's output - https://paste.fedoraproject.org/paste/Iz4ix5OO~wHiQFfpJenFHQ. I've come across this because I change the line `self.mount_a.umount_wait()` to `self.mount_a.umount_wait(require_clean=True)`. Otherwise, the test was working fine. 

Also, the `ceph-fuse` process is shown as defunct when listed by `ps`. But, I guess that is only because its parent process (the python test) is still alive,

batrick commented:
@rishabh-d-dave the issue was that the client is waitign for the MDS to respond that the session is closed but the MDS drops these messages without replying:

https://github.com/rishabh-d-dave/ceph/blob/717630207a958aa4e4d598a05dbd4b86678acfc1/src/mds/Server.cc#L423-L429

I've changed the behavior for the MDS and added a commit to this PR. Also, please consider these changes to your QA test:

````
diff --git a/qa/tasks/cephfs/fuse_mount.py b/qa/tasks/cephfs/fuse_mount.py
index 3aa07a339e7..57f25ede489 100644
--- a/qa/tasks/cephfs/fuse_mount.py
+++ b/qa/tasks/cephfs/fuse_mount.py
@@ -261,7 +261,7 @@ class FuseMount(CephFSMount):
         assert not self.is_mounted()
         self._fuse_conn = None
 
-    def umount_wait(self, force=False, require_clean=False):
+    def umount_wait(self, force=False, require_clean=False, timeout=900):
         """
         :param force: Complete cleanly even if the MDS is offline
         """
@@ -282,7 +282,7 @@ class FuseMount(CephFSMount):
         try:
             if self.fuse_daemon:
                 # Permit a timeout, so that we do not block forever
-                run.wait([self.fuse_daemon], 900)
+                run.wait([self.fuse_daemon], timeout)
         except MaxWhileTries:
             log.error("process failed to terminate after unmount.  This probably"
                       "indicates a bug within ceph-fuse.")
diff --git a/qa/tasks/cephfs/test_misc.py b/qa/tasks/cephfs/test_misc.py
index b13e56fa119..3956c457091 100644
--- a/qa/tasks/cephfs/test_misc.py
+++ b/qa/tasks/cephfs/test_misc.py
@@ -133,7 +133,6 @@ class TestMisc(CephFSTestCase):
     def test_unmount_for_evicted_client(self):
         """Test if client hangs on unmount after evicting the client."""
         # Only one client is required for this test.
-        self.mount_b.umount()
         self.mount_b.umount_wait()
 
         self.fs.mon_manager.raw_cluster_cmd('fs', 'set', self.fs.name,
@@ -150,8 +149,7 @@ class TestMisc(CephFSTestCase):
                                             'id={0}'.format(client_id))
 
         # Test: Instead of hanging, client should unmount.
-        self.mount_a.umount()
-        self.mount_a.umount_wait()
+        self.mount_a.umount_wait(timeout=30)
 
     def test_filtered_df(self):
         pool_name = self.fs.get_data_pool_name()
````

batrick commented:
@ukernel please review the MDS bits.

rishabh-d-dave commented:
> @rishabh-d-dave the issue was that the client is waitign for the MDS to respond that the session is closed but the MDS drops these messages without replying:
>
> https://github.com/rishabh-d-dave/ceph/blob/717630207a958aa4e4d598a05dbd4b86678acfc1/src/mds/Server.cc#L423-L429
>
> I've changed the behavior for the MDS and added a commit to this PR. Also, please consider these changes to your QA test:

@batrick I still see ceph-fuse exiting with return value -6 and the same traceback (I still pass the argument `require_clean=True` to `umount_wait()`.

The test  after its completion leaves the cluster running. So, I did the test myself (mounting, attaching gdb, evicting and unmounting client) . There were no tracebacks and `ceph-fuse` exited normally.

I've tried this too: instead of letting the test run completely, I _pause_ it after the `max_mds` is set to 2 for the FS and I conduct the test manually while the the automated test is still paused and then allow the automated test to complete. The results are the same: client created manually works fine but for the automated test it exits with -6.

With respect to the traceback, the issues seems to be with this line [1]; commenting it off fixes the issue. I don't know what role is played by `waiting_for_map`, but I'll start taking a look in there next.

[1] https://github.com/ceph/ceph/blob/master/src/osdc/Objecter.cc#L5009

ukernel commented on src/client/Client.cc:
```diff
@@ -5964,6 +5966,16 @@ void Client::tick()
+      mono_time_curr_req_stamp - mono_clock::zero());
+    if (mounted && !mds_requests.empty() && ticks_past_since_req.count() >
+	MAX_TIME_FOR_A_REQ)
+      objecter->get_latest_version(0, int_curr_req_stamp.count(), NULL);
```
the second argument of get_latest_version() is epoch_t. why stamp here?

ukernel replied:
I think it's better to call Objecter::maybe_request_map() here.

batrick replied:
+1 `Objecter::maybe_request_map()`

@rishabh-d-dave

batrick replied:
@rishabh-d-dave @ukernel's suggested change fixes your issue.

The problem was an assertion in the client:

````
     0> 2018-04-17 18:09:34.172 7f79c43cac00 -1 /home/pdonnell/ceph/src/osdc/Objecter.cc: In function 'virtual Objecter::~Objecter()' thread 7f79c43cac00 time 2018-04-17 18:09:34.173531
/home/pdonnell/ceph/src/osdc/Objecter.cc: 5008: FAILED assert(waiting_for_map.empty())

 ceph version 13.0.2-975-ga749882 (a74988218e522d6ecf9b98289539e0c615c14d56) mimic (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x102) [0x7f79bb33fb42]
 2: (()+0x292d17) [0x7f79bb33fd17]
 3: (Objecter::~Objecter()+0x158f) [0x7f79bb660b2f]
 4: (Objecter::~Objecter()+0x9) [0x7f79bb660b89]
 5: (StandaloneClient::~StandaloneClient()+0x2e) [0x866055787e]
 6: (StandaloneClient::~StandaloneClient()+0x9) [0x86605578b9]
 7: (main()+0x586) [0x86604dffd6]
 8: (__libc_start_main()+0xf0) [0x7f79ba03a830]
 9: (_start()+0x29) [0x86604e24a9]
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.

````

An abort signal (`SIGABRT`) is returned as a `-6` return code by Python's subprocess library.

rishabh-d-dave replied:
@batrick @ukernel Done.

batrick commented on qa/tasks/cephfs/test_client_recovery.py:
```diff
@@ -476,3 +476,23 @@ def _test_fsync(self, dirfsync):
+        """Test if client hangs on unmount after evicting the client."""
+        self.fs.mon_manager.raw_cluster_cmd('fs', 'set', self.fs.name,
+                                            'allow_multimds', 'true',
+                                            '--yes-i-really-mean-it')
```
remove allow_multimds call

rishabh-d-dave replied:
@batrick Done.

batrick commented on qa/tasks/cephfs/test_client_recovery.py:
```diff
@@ -476,3 +476,23 @@ def _test_fsync(self, dirfsync):
+        self.fs.mon_manager.raw_cluster_cmd('fs', 'set', self.fs.name,
+                                            'allow_multimds', 'true',
+                                            '--yes-i-really-mean-it')
+        self.fs.set_max_mds(2)
```
Why do we need 2 active MDS?

rishabh-d-dave replied:
I was following the recipe provided on the tracker, but yeah, it's not required. Fixed.

taodd commented:
@ukernel, @batrick @rishabh-d-dave why can not we just request for new osdmap in Client::ms_handle_remote_reset ??
the solution here would still make the unmount may wait for 120 s ?

rishabh-d-dave commented on src/client/Client.cc:
```diff
@@ -5964,6 +5966,16 @@ void Client::tick()
   }
 
+  {
+    chrono::duration<double, std::milli> ticks_past_since_req =
```
I don't know when and how I made the mistake here (this will make client wait in milliseconds); found it while testing locally after replacing `get_latest_version()` by `maybe_request_map()`. I'll fix in the upcoming push.

rishabh-d-dave commented on src/mds/Server.cc:
```diff
@@ -420,12 +420,11 @@ void Server::handle_client_session(MClientSession *m)
+        dout(10) << "session alredy closed|closing|killing, responding with CEPH_SESSION_CLOSE" << dendl;
+        mds->send_message_client(new MClientSession(CEPH_SESSION_CLOSE), session);
+        m->put();
+        return;
```
I was coming across few issues to which Patrick found this solution. You can refer the following comments -
https://github.com/ceph/ceph/pull/21065#issuecomment-380389137
https://github.com/ceph/ceph/pull/21065#issuecomment-380907455

taodd replied:
Here the session is already killed or killing because of the eviction, so i don't think it's properly for the mds to reply to the evicted client here. we could just let the hang request be cleaned after the client realize it is in that blacklist.

batrick replied:
@taodd That would be possible but I don't see a downside to the MDS telling the client to close its session.

@ukernel do you see a problem with this?

ukernel replied:
when client requests session close, it should have no dirty caps. It should be harmless that mds blindly sends a reply.  kernel client handles this case by using timed wait. I personally prefer kernel's appocah

rishabh-d-dave commented:
> Please redo this PR with these changes and test yourself to confirm.

They do, as I've mentioned in the mail. I've retested and updated the PR.

> Sorry to send you on a goosechase but I think the process was educational for everyone, even me.

No problem.  Thanks for all the help!  :)
